## Objective

The objective of this notebook is to prepare the cleaned dataset for machine learning by:
- Selecting relevant features
- Encoding categorical variables
- Splitting data into training and testing sets
- Applying feature scaling
- Training baseline machine learning models to predict final grade (G3)


In [2]:
import pandas as pd


In [3]:
df = pd.read_csv("../data/cleaned_student_data.csv")
df.head()


,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,Dalc,Walc,health,absences,G1,G2,G3,total_alcohol,avg_grade,absence_level
0,GP,0,18,U,GT3,0,4,4,at_home,teacher,...,1,1,3,6,5,6,6,2,5.666667,Moderate
1,GP,0,17,U,GT3,1,1,1,at_home,other,...,1,1,3,4,5,5,6,2,5.333333,Low
2,GP,0,15,U,LE3,1,1,1,at_home,other,...,2,3,3,10,7,8,10,5,8.333333,Moderate
3,GP,0,15,U,GT3,1,4,2,health,services,...,1,1,5,2,15,14,15,2,14.666667,Low
4,GP,0,16,U,GT3,1,3,3,other,other,...,1,2,5,4,6,10,10,3,8.666667,Low


In [4]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score,accuracy_score,confusion_matrix

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor


In [5]:
target="G3"

**Targeted variable for our modelling is G3**

In [6]:
features=["G1","G2","studytime","failures","famsup","schoolsup","health","freetime",
          "Dalc","Walc","paid","internet","famsup","schoolsup","activities"]

**Assigning variables as X AND Y for features and target from our data set👇🏻**

In [7]:
X=df[features]
Y=df[target]

In [8]:
X

,G1,G2,studytime,failures,famsup,schoolsup,health,freetime,Dalc,Walc,paid,internet,famsup,schoolsup,activities
0,5,6,2,0,0,1,3,3,1,1,0,0,0,1,0
1,5,5,2,0,1,0,3,3,1,1,0,1,1,0,0
2,7,8,2,3,0,1,3,3,2,3,1,1,0,1,0
3,15,14,3,0,1,0,5,2,1,1,1,1,1,0,1
4,6,10,2,0,1,0,5,3,1,2,1,0,1,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,9,9,2,2,1,0,4,5,4,5,1,0,1,0,0
391,14,16,1,0,0,0,2,4,3,4,0,1,0,0,0
392,10,8,1,3,0,0,3,5,3,3,0,0,0,0,0
393,11,12,1,0,0,0,5,4,3,4,0,1,0,0,0


NOW ENCODING THE BINARY VALUES INTO A NUMERICAL ONE👇🏻

In [9]:
X=pd.get_dummies(X,drop_first=True)
X.head()

,G1,G2,studytime,failures,famsup,schoolsup,health,freetime,Dalc,Walc,paid,internet,famsup,schoolsup,activities
0,5,6,2,0,0,1,3,3,1,1,0,0,0,1,0
1,5,5,2,0,1,0,3,3,1,1,0,1,1,0,0
2,7,8,2,3,0,1,3,3,2,3,1,1,0,1,0
3,15,14,3,0,1,0,5,2,1,1,1,1,1,0,1
4,6,10,2,0,1,0,5,3,1,2,1,0,1,0,0


## Train-Test Split

The dataset is split into training and testing sets to evaluate model generalization.


In [10]:
X_TRAIN,X_TEST,Y_TRAIN,Y_TEST=train_test_split(X,Y,random_state=42,test_size=0.2)

In [11]:
X_TRAIN.head()

,G1,G2,studytime,failures,famsup,schoolsup,health,freetime,Dalc,Walc,paid,internet,famsup,schoolsup,activities
181,12,13,2,0,0,0,3,2,1,2,1,1,0,0,1
194,13,14,1,0,0,0,3,3,1,1,0,1,0,0,0
173,8,7,2,3,0,0,3,3,1,1,0,1,0,0,1
63,10,9,3,0,1,1,4,4,2,4,1,1,1,1,1
253,8,9,1,0,0,0,3,3,1,3,0,0,0,0,1


In [12]:
X_TEST.head()

,G1,G2,studytime,failures,famsup,schoolsup,health,freetime,Dalc,Walc,paid,internet,famsup,schoolsup,activities
78,8,8,1,3,1,1,3,5,1,1,0,1,1,1,1
371,14,12,1,0,1,0,3,3,2,3,1,1,1,0,1
248,3,5,2,1,1,0,5,3,1,3,0,1,1,0,0
55,8,9,2,0,0,0,2,3,1,1,1,1,0,0,1
390,9,9,2,2,1,0,4,5,4,5,1,0,1,0,0


In [13]:
Y_TEST.head()

78     10
371    12
248     5
55     10
390     9
Name: G3, dtype: int64

## Feature Scaling

Numerical features are scaled to ensure fair contribution during model training.


In [14]:
scalar=StandardScaler()
X_trained_scaled=scalar.fit_transform(X_TRAIN)

In [15]:
X_trained_scaled

array([[ 0.33205033,  0.62616324, -0.05685286, ..., -1.23612297,
        -0.40219983,  1.00634927],
       [ 0.64340909,  0.89283114, -1.25455313, ..., -1.23612297,
        -0.40219983, -0.99369079],
       [-0.91338472, -0.97384417, -0.05685286, ..., -1.23612297,
        -0.40219983,  1.00634927],
       ...,
       [-0.60202596, -0.44050837, -0.05685286, ...,  0.808981  ,
        -0.40219983,  1.00634927],
       [ 0.64340909,  1.15949904,  1.14084741, ...,  0.808981  ,
        -0.40219983,  1.00634927],
       [-0.2906672 ,  0.62616324, -1.25455313, ...,  0.808981  ,
        -0.40219983,  1.00634927]], shape=(316, 15))

In [16]:
X_test_scaled=scalar.transform(X_TEST)

## Model 1: Linear Regression

Linear Regression is used as a baseline model to understand linear relationships between features and target.


In [17]:
LR=LinearRegression()
Model_fit=LR.fit(X_trained_scaled,Y_TRAIN)
y_pred=LR.predict(X_test_scaled)


In [18]:
y_pred

array([ 6.57732526, 12.05540638,  3.58856253,  7.4536864 ,  8.20091065,
       13.1170006 , 18.71919873,  6.84143181,  6.80610622, 12.21934383,
       15.04609878,  5.75850463, 14.33885206, 12.1071138 , 13.89373163,
        8.00606498,  6.43981533, 10.95431382, 15.53963299,  8.05549856,
       14.53749825, 16.30079545, 14.96646783,  5.54029533,  8.82513159,
       19.91436721, 10.83590324,  8.40991248, 17.14386243,  9.89843444,
        7.99855773,  7.37442174, 15.52246564, 12.73175332,  6.61960491,
        3.46797637, -2.38249504, 14.79017844, 11.4566706 ,  7.54356391,
        6.00068646, 10.68886695, 14.53967421,  7.03920704, 15.80900601,
        9.12805587, 12.35769266, 13.58035753, 12.17299733, 15.93960881,
       13.63897529, 14.53033309, 10.92788065,  7.78245441,  3.81442229,
       13.22830175, 10.12751162,  5.88966531, 15.41239844, 16.4016139 ,
       12.18163495,  8.17012731,  6.48936044,  3.94167422,  4.47808991,
       17.50906475,  7.59120579,  9.68431727,  9.02680219, 15.33

In [19]:
lr_mse=mean_squared_error(Y_TEST,y_pred)
lr_rmse=rmse=np.sqrt(lr_mse)
lr_mae=mean_absolute_error(Y_TEST,y_pred)
r2_lr=r2_score(Y_TEST,y_pred)
print("MSE:",lr_mse)
print("RMSE:",lr_rmse)
print("MAE:",lr_mae)
print("R2 SCORE IS:",r2_lr)


MSE: 4.826395315013389
RMSE: 2.1969058502843013
MAE: 1.4040056487516746
R2 SCORE IS: 0.7646240336870678


## Model 2: Decision Tree Regressor

Decision Trees capture non-linear relationships and feature interactions.


In [20]:
dt=DecisionTreeRegressor(max_depth=5,
    random_state=42)
dt.fit(X_TRAIN,Y_TRAIN)
y_pred_dt=dt.predict(X_TEST)


In [21]:
dt_mse=mean_squared_error(Y_TEST,y_pred_dt)
dt_rmse=np.sqrt(dt_mse)
dt_mae=mean_absolute_error(Y_TEST,y_pred_dt)
r2_dt=r2_score(Y_TEST,y_pred_dt)
print("MSE:",dt_mse)
print("RMSE:",dt_rmse)
print("MAE:",dt_mae)
print("R2 SCORE IS:",r2_dt)

MSE: 4.481697138280708
RMSE: 2.117001922124944
MAE: 1.3893145070193702
R2 SCORE IS: 0.78143444003368


## Model 3: Random Forest Regressor

Random Forest improves performance by combining multiple decision trees.


In [22]:
rfr=RandomForestRegressor(n_estimators=100,random_state=42,max_depth=8)
model_fitting=rfr.fit(X_TRAIN,Y_TRAIN)
y_pred_rf=model_fitting.predict(X_TEST)

In [23]:
print("Random Forest Performance:")
rf_mse=mean_squared_error(Y_TEST,y_pred_rf)
rf_rmse=np.sqrt(rf_mse)
rf_mae=mean_absolute_error(Y_TEST,y_pred_rf)
r2_rf=r2_score(Y_TEST,y_pred_rf)
print("MSE:",rf_mse)
print("RMSE:",rf_rmse)
print("MAE:",rf_mae)
print("R2 SCORE IS:",r2_rf)



Random Forest Performance:
MSE: 5.768051583170163
RMSE: 2.401676827379188
MAE: 1.5218830634084852
R2 SCORE IS: 0.7187008882367629


In [24]:
from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(
    RandomForestRegressor(
        n_estimators=200,
        max_depth=8,
        random_state=42
    ),
    X, Y,
    scoring='r2',
    cv=5
)

print("Cross-validated R2 scores:", cv_scores)
print("Mean R2:", cv_scores.mean())



Cross-validated R2 scores: [0.75508018 0.884213   0.79172297 0.75549605 0.67506293]
Mean R2: 0.7723150270572728


Cross-validation was performed to obtain a stable estimate of model performance.
This reduces dependency on a single train-test split.


In [25]:
import pandas as pd

results = pd.DataFrame({
    "Model": ["Linear Regression", "Decision Tree", "Random Forest"],
    "MAE": [lr_mae, dt_mae, rf_mae],
    "RMSE": [lr_rmse,dt_rmse, rf_rmse],
    "R2": [r2_lr, r2_dt, r2_rf]
})

results


,Model,MAE,RMSE,R2
0,Linear Regression,1.404006,2.196906,0.764624
1,Decision Tree,1.389315,2.117002,0.781434
2,Random Forest,1.521883,2.401677,0.718701


## Final Model Comparison and Conclusion

Three regression models were trained to predict the final student grade (G3):

- Linear Regression
- Decision Tree Regressor
- Random Forest Regressor

### Model Performance Summary

| Model              | MAE ↓ | RMSE ↓ | R² ↑ |
|-------------------|-------|--------|------|
| Linear Regression | ~1.40 | ~2.19  | ~0.76 |
| Decision Tree     | ~1.39 | ~2.11  | ~0.78 |
| Random Forest     | ~1.52 | ~2.40  | ~0.71 |

### Key Observations

- **Decision Tree Regressor achieved the highest R² score**, indicating the best overall fit on test data.
- **Linear Regression performed competitively**, suggesting a strong linear relationship between predictors (especially G1, G2) and the target variable G3.
- **Random Forest underperformed** compared to expectations due to the relatively small dataset and dominant linear signals from prior grades.

### Important Notes

- Accuracy metric was intentionally excluded as it is not suitable for regression problems.
- Variations in Decision Tree performance are expected due to its high variance nature.
- Prior academic scores (G1, G2) are the strongest predictors of final performance.

### Final Conclusion

The Decision Tree Regressor is selected as the best-performing model for this dataset, while Linear Regression remains a strong and interpretable baseline. The modeling results confirm that historical academic performance is the most influential factor in predicting final grades.

Notebook-05 successfully completes model training, evaluation, and comparison.  
The project is now ready for final reporting and insights presentation.


## loading for pkl file 
- changes by ramesh

1.joblib.dump(dt, "student_model.pkl")

Saves the trained Decision Tree Regressor (dt) into a file named
student_model.pkl.

This converts the in-memory model object into a persistent file on disk.

The saved file can later be loaded by Flask for prediction without retraining.

2.joblib.dump(X.columns.tolist(), "model_features.pkl")

Saves the exact feature names and their order used during training.

This ensures that user input in Flask is processed in the same format as training data.


In [28]:
import joblib

joblib.dump(dt, "student_model.pkl")
joblib.dump(X.columns.tolist(), "model_features.pkl")

print("Model and features saved in notebook folder")


Model and features saved in notebook folder
